### PR TITLE
Add optimizations introduced in MLSys paper

### PR DESCRIPTION
**Pull request by Xiuyu-Li**
In this PR we implement optimization techniques introduced in our MLSys paper "TorchSparse: Efficient Point Cloud Inference Engine". These include adaptive matrix multiplication grouping and vectorized, quantized and fused locality-aware memory access. More details are described in the [paper](https://torchsparse.mit.edu/assets/paper.pdf).

### CODE REVIEW

File: torchsparse/backend/convolution/convolution_cuda.cu [hunk 1]
```diff
@@ -10,30 +10,246 @@
...
+  int index = blockIdx.x * blockDim.x + threadIdx.x;
+  int i, j;
+  int offset = (sizeof(scalar_t) == 4) ? 2 : 3;
+  i = index / (c >> offset);
+  j = index % (c >> offset);
+  if (i >= n) return;
+  int4 tmps[1];
+  // tmps[0].x = tmps[0].y = tmps[0].z = tmps[0].w = 0;
+  CONVERT_INT4(tmps) = CONVERT_INT4(in_feat[i * c + (j << offset)]);
+  // float verify = *(reinterpret_cast<float*>(tmps) + 3);
```
Comment (kentangSJTU) [inline torchsparse/backend/convolution/convolution_cuda.cu:86]: We can remove those comments for debugging (all `printf` and `std::cout`s).

File: torchsparse/backend/convolution/convolution_cuda.cu [hunk 2]
```diff
@@ -52,8 +268,410 @@ __global__ void scatter_kernel(const int n_in, const int n_out, const int c,
...
+
+  at::Tensor neighbor_offset_gpu = neighbor_offset.to(in_feat.device());
+  at::Tensor neighbor_offset_cum_gpu =
+      torch::cumsum(neighbor_offset_gpu, 0).to(at::ScalarType::Int);
+  at::Tensor neighbor_offset_cum =
+      neighbor_offset_cum_gpu.to(neighbor_offset.device());
+  // at::Tensor neighbor_map_cpu = neighbor_map.to(neighbor_offset.device());
+  at::Tensor cum_buffer_sizes = torch::zeros_like(neighbor_offset);
+
+  // int tile_size = 1024;
```
Comment (kentangSJTU) [inline torchsparse/backend/convolution/convolution_cuda.cu:418]: We can remove this line and `tile_id` (which are not used).

File: torchsparse/backend/convolution/convolution_cuda.cu [hunk 3]
```diff
@@ -52,8 +268,410 @@ __global__ void scatter_kernel(const int n_in, const int n_out, const int c,
...
+          -1);
+      kernel = torch::cat(
+          {kernel, torch::zeros({kernel.size(0), 8 - (kernel.size(1) % 8),
+                                 kernel.size(2)},
+                                options)},
+          1);
+    }
+    if (out_feat.size(1) % 8 != 0) {
+      out_feat = torch::cat(
+          {out_feat, torch::zeros({out_feat.size(0), 8 - (in_feat.size(1) % 8)},
```
Comment (kentangSJTU) [inline torchsparse/backend/convolution/convolution_cuda.cu:441]: The format of line 441-470 is a little bit messy. Is it caused by the linting tool?

File: torchsparse/backend/convolution/convolution_cuda.cu [hunk 4]
```diff
@@ -52,8 +268,410 @@ __global__ void scatter_kernel(const int n_in, const int n_out, const int c,
...
+          {out_feat, torch::zeros({out_feat.size(0), 4 - (in_feat.size(1) % 4)},
+                                  options)},
+          -1);
+      kernel = torch::cat({kernel, torch::zeros({kernel.size(0), kernel.size(1),
+                                                 4 - (kernel.size(2) % 4)},
+                                                options)},
+                          -1);
+    }
+  }
+  // std::cout << in_feat.size(0) << " " << in_feat.size(1) << " " <<
```
Comment (kentangSJTU) [inline torchsparse/backend/convolution/convolution_cuda.cu:472]: We can remove the debugging code.

File: torchsparse/backend/convolution/convolution_cuda.cu [hunk 5]
```diff
@@ -52,8 +268,410 @@ __global__ void scatter_kernel(const int n_in, const int n_out, const int c,
...
+        *std::max_element(neighbor_offset.data_ptr<int>(),
+                          neighbor_offset.data_ptr<int>() + kernel_volume / 2);
+    max_kmap_size =
+        std::max(max_kmap_size,
+                 *std::max_element(
+                     neighbor_offset.data_ptr<int>() + kernel_volume / 2 + 1,
+                     neighbor_offset.data_ptr<int>() + kernel_volume));
+    max_kmap_size = std::max(max_kmap_size, 1);
+    // (N, c) X (c, o) = (N, o)
+    // torch::mm_out(out_feat, in_feat, kernel[mid_kernel]);
```
Comment (kentangSJTU) [inline torchsparse/backend/convolution/convolution_cuda.cu:501]: We can remove line 500-501 to avoid possible confusion. Currently we are doing the computation for kernel offset = (0, 0, 0) in the last, instead of in the first.

File: torchsparse/backend/convolution/convolution_cuda.cu [hunk 6]
```diff
@@ -52,8 +268,410 @@ __global__ void scatter_kernel(const int n_in, const int n_out, const int c,
...
+      }));
+
+  at::Tensor in_buffer_activated, out_buffer_activated, kernel_buffer;
+  int buffer_st;
+  int cur_buffer_size;
+  // for each group
+  // mm_ops = 0, sep; mm_ops = 1, BMM
+  int kernel_cnt = 0;
+  for (int i = 0; i < groups.size(); i++) {
+    // case 0 and 1: no padding
```
Comment (kentangSJTU) [inline torchsparse/backend/convolution/convolution_cuda.cu:565]: Let's remove this comment, it is related to our old implementation and is confusing for the current one (`mm_ops=1` usually means that there are paddings).

File: torchsparse/backend/convolution/convolution_cuda.cu [hunk 7]
```diff
@@ -85,7 +703,8 @@ void convolution_forward_cuda(at::Tensor in_feat, at::Tensor out_feat,
     in_buffer_size = std::max(in_buffer_size, 1);
 
     // (N, c) X (c, o) = (N, o)
-    torch::mm_out(out_feat, in_feat, kernel[mid_kernel]);
+    int mid_kmap_idx = conv_mode != 2 ? kernel_volume / 2 : kernel_volume - 1;
```
Comment (kentangSJTU) [inline torchsparse/backend/convolution/convolution_cuda.cu:706]: Let's add a comment explaining `conv_mode=2` means reordering the kernel.

File: torchsparse/nn/functional/build_kmap.py [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+__all__ = ['build_kernel_map']
+
+
+def build_kernel_map(_coords: torch.Tensor,
+                     kernel_size: Union[int, Tuple[int, ...]] = 2,
+                     stride: Union[int, Tuple[int, ...]] = 2,
+                     tensor_stride: Union[int, Tuple[int, ...]] = 1,
+                     mode='hashmap') -> torch.Tensor:
+    # print(_coords.shape, kernel_size, stride, tensor_stride)
```
Comment (kentangSJTU) [inline torchsparse/nn/functional/build_kmap.py:18]: We can remove the debugging code.

File: torchsparse/nn/functional/conv.py [hunk 9]
```diff
@@ -17,189 +16,190 @@ class ConvolutionFunction(Function):
...
+        output = SparseTensor(coords=coords,
+                              feats=feats,
+                              stride=input.stride,
+                              buffer=input.buffer)
     elif not transposed:
-        output_stride = tuple(input.stride[k] * stride[k] for k in range(3))
+        kmap = input.kmaps.get((input.stride, kernel_size, stride, dilation))
+        if kmap is None:
+            if any(s > 1 for s in stride):
+                # print('before', coords.shape)
```
Comment (kentangSJTU) [inline torchsparse/nn/functional/conv.py:142]: We can remove the debugging code (all `print`s).

File: torchsparse/nn/modules/conv.py [hunk 10]
```diff
@@ -21,7 +21,8 @@ def __init__(self,
                  stride: Union[int, Tuple[int, ...]] = 1,
                  dilation: int = 1,
                  bias: bool = False,
-                 transposed: bool = False) -> None:
+                 transposed: bool = False,
+                 config=None) -> None:
```
Comment (kentangSJTU) [inline torchsparse/nn/modules/conv.py:25]: Let's add some typing here. `config` should be `Dict`.

File: torchsparse/utils/reorder.py [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+def reorder_kernel(net):
+    for child_name, child in net.named_children():
+        if isinstance(child, spnn.Conv3d):
+            if (len(child.kernel.data.shape) == 3):
+                kernel_volume = child.kernel.data.size(0)
+                kernel_cont = torch.zeros_like(child.kernel.data)
+                ind = 0
+                while ind < kernel_volume - 1:
+                    kernel_cont[ind] = child.kernel.data[ind // 2]
+                    kernel_cont[ind + 1] = child.kernel.data[kernel_volume - 1
```
Comment (kentangSJTU) [inline torchsparse/utils/reorder.py:17]: Let's check the formatting for lines 17-18, 21-22.

**Review by kentangSJTU — COMMENTED** [review]
Hi @Xiuyu-Li, thank you for your great efforts in organizing the code release! I've been through the current implementation and I have some small comments related to the formatting of the code. Would you mind having a look at them and resolving those conversations? Thank you!

File: torchsparse/backend/convolution/convolution_cuda.cu (hunk 3 above)
Comment (Xiuyu-Li) [reply torchsparse/backend/convolution/convolution_cuda.cu:446]: Yes. The clang-format seems to be pretty insistent on this formating.

**Review by kentangSJTU — APPROVED** [review]

File: torchsparse/nn/modules/bev.py [hunk 12]
```diff
@@ -86,12 +86,11 @@ def forward(self, input: SparseTensor) -> torch.Tensor:
         coords, feats, stride = input.coords, input.feats, input.stride
```
Comment (zhijian-liu) [inline torchsparse/nn/modules/bev.py:86]: @Xiuyu-Li, please revert these changes. These are required for proper backward support.

**Comment by DableUTeeF** [comment]
This introduces `cached_property` which doesn't exist in python 3.7 or below.

Some instructions, for example which specific version worked with what python version is really needed.

**Comment by Xiuyu-Li** [comment]
@DableUTeeF Thanks for bringing this out. We will update the code / README and upload a documentation soon. For now, you can fix this by first running `pip install cached-property`, then changing https://github.com/mit-han-lab/torchsparse/blob/36ca82457a053207ecb31578badb887e8356a482/torchsparse/nn/modules/conv.py#L2 to `from cached_property import cached_property`.
